pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -624,6 +627,83 @@ public Object visit(ASTType node, Object data) {
         return data;
     }
 
+    private void populateVariableDeclaratorFromType(ASTLocalVariableDeclaration node) {
+        ASTType type = node.getTypeNode();
+        // also assign this type to VariableDeclarator and VariableDeclaratorId
+        TypeNode var = node.getFirstChildOfType(ASTVariableDeclarator.class);
+        if (var != null) {
+            var.setTypeDefinition(type.getTypeDefinition());
+            var = var.getFirstChildOfType(ASTVariableDeclaratorId.class);
+        }
+        if (var != null) {
+            var.setTypeDefinition(type.getTypeDefinition());
+        }
+    }
+
+    @Override
+    public Object visit(ASTLocalVariableDeclaration node, Object data) {
+        super.visit(node, data);
+        // resolve "var" types: Upward projection of the type of the initializer expression
+        ASTType type = node.getFirstChildOfType(ASTType.class);
+        if (type != null && type.isTypeInferred()) {
+            ASTVariableInitializer initializer = node.getFirstDescendantOfType(ASTVariableInitializer.class);
+            if (initializer != null && initializer.jjtGetChild(0) instanceof ASTExpression) {
+                // only Expression is allowed, ArrayInitializer is not allowed in combination with "var".
+                ASTExpression expression = (ASTExpression) initializer.jjtGetChild(0);
+                type.setTypeDefinition(expression.getTypeDefinition());
```

Comment:
Not having a Type node in that case would also avoid juggling with the inferred time, from the Expression => Type => VariableDeclarator => VariableDeclaratorId